### PR TITLE
Emit warning when bound is trivially false

### PR DESCRIPTION
**Pull request by compiler-errors**
We really should be rejecting bounds that are global, even if they have higher-ranked lifetimes. We have the ability to prove whether `for<'a> &'a mut String: Clone` is true or not, so there's no reason to silently allow this to typecheck, especially because it's inconsistent with a non-HRTB like `&'static mut String: Clone` and is the source of ICEs.

For now, though, just add a warning, so that when an ICE _does_ occur (because sometimes it will), then the user knows what might've caused it.

This is a revival of #95611.

### CODE REVIEW

**Comment by compiler-errors** [comment]
r? types

**Comment by jackh726** [comment]
Let's specifically r? @nikomatsakis, since he has some specific thoughts on this :)

(Though I personally do think a warning is good)

File: src/test/ui/associated-types/issue-69398.stderr [hunk 1]
```diff
@@ -0,0 +1,14 @@
+warning: the where-clause bound `<T as Broken>::Assoc: Foo` is impossible to satisfy
```
Comment (jackh726) [inline src/test/ui/associated-types/issue-69398.stderr:1]: It's a bit unfortunate that the bound we say is not the bound that's written.

File: src/test/ui/consts/issue-67696-const-prop-ice.stderr [hunk 2]
```diff
@@ -0,0 +1,26 @@
+warning: the where-clause bound `[for<'r> fn(&'r ())]: Copy` is impossible to satisfy
```
Comment (jackh726) [inline src/test/ui/consts/issue-67696-const-prop-ice.stderr:1]: This is certainly not great.

**Review by jackh726 — COMMENTED** [review]
Warnings here...aren't great.

**Comment by QuineDot** [comment]
> We really should be rejecting bounds that are global, even if they have higher-ranked lifetimes.

[As per RFC 2056,](https://github.com/rust-lang/rust/issues/48214) they should not be rejected.  The HRTB's this targets [can currently be used to achieve trivial bounds without `#[feature(trivial_bounds)]`](https://github.com/rust-lang/rust/issues/48214#issuecomment-1150463333). The motivating example in that comment is [the inability to omit unsatisfiable trait members](https://github.com/rust-lang/rfcs/issues/2829), i.e. it works around a practical problem.  Until true trivial bounds are allowed on stable, it's arguably a more sensible workaround than [leaving the bound off entirely](https://github.com/rust-lang/rfcs/pull/3245#issuecomment-1120097693) -- especially once such a refined implementation can in other cases be called / relied upon.

Warning makes sense though.

**Pull request closed without merging**